modules/caddypki/command.go

```diff
@@ -30,69 +32,153 @@ func init() {
+This command can be used in one of two ways. Either by specifying
+which certificate to untrust by a direct path to the certificate
+file with the --cert flag, or by fetching the root certificate for
+the CA from the admin API (default behaviour).
 
-If no flags are specified, --ca=local is assumed.`,
+If the admin API is used, then the CA defaults to 'local'. You may
+specify the ID of another CA with the --ca flag. By default, this
+will attempt to connect to the Caddy's admin API running at
+'` + caddy.DefaultAdminListen + `' to fetch the root certificate.
+You may explicitly specify the --address, or use the --config flag
+to load the admin address from your config, if not using the default.`,
 		Flags: func() *flag.FlagSet {
 			fs := flag.NewFlagSet("untrust", flag.ExitOnError)
-			fs.String("ca", "", "The ID of the CA to untrust")
 			fs.String("cert", "", "The path to the CA certificate to untrust")
+			fs.String("ca", "", "The ID of the CA to untrust (defaults to 'local')")
+			fs.String("address", "", "Address of the administration API listener (if --config is not used)")
+			fs.String("config", "", "Configuration file (if --address is not used)")
+			fs.String("adapter", "", "Name of config adapter to apply (if --config is used)")
 			return fs
 		}(),
 	})
 }
 
-func cmdTrust(fs caddycmd.Flags) (int, error) {
-	// we have to create a sort of dummy context so that
-	// the CA can provision itself...
-	ctx, cancel := caddy.NewContext(caddy.Context{Context: context.Background()})
-	defer cancel()
+func cmdTrust(fl caddycmd.Flags) (int, error) {
+	caID := fl.String("ca")
+	addrFlag := fl.String("address")
+	configFlag := fl.String("config")
+	configAdapterFlag := fl.String("adapter")
 
-	// provision the CA, which generates and stores a root
-	// certificate if one doesn't already exist in storage
-	ca := CA{
-		storage: caddy.DefaultStorage,
+	// Prepare the URI to the admin endpoint
+	if caID == "" {
+		caID = DefaultCAID
+	}
+	uri := amdinPKICertificatesEndpoint + caID
+
+	// Prefer the address flag, but if a config is specified,
+	// try to load the admin address from there.
+	adminAddr := addrFlag
+	if adminAddr == "" && configFlag != "" {
+		// get the config in caddy's native format
+		config, configFile, err := caddycmd.LoadConfig(configFlag, configAdapterFlag)
+		if err != nil {
+			return caddy.ExitCodeFailedStartup, err
+		}
+		if configFile == "" {
+			return caddy.ExitCodeFailedStartup, fmt.Errorf("no config file to load")
+		}
+
+		// get the address of the admin listener
+		if len(config) > 0 {
+			var tmpStruct struct {
+				Admin caddy.AdminConfig `json:"admin"`
+			}
+			err = json.Unmarshal(config, &tmpStruct)
+			if err != nil {
+				return caddy.ExitCodeFailedStartup,
+					fmt.Errorf("unmarshaling admin listener address from config: %v", err)
+			}
+			adminAddr = tmpStruct.Admin.Listen
+		}
 	}
-	err := ca.Provision(ctx, DefaultCAID, caddy.Log())
+	if adminAddr == "" {
+		adminAddr = caddy.DefaultAdminListen
+	}
```

Comment:
Yikes, this is complicated. (I know it's not new.)

If we're going to be exporting the function to perform API requests, I think the address of the admin endpoint should be irrelevant to the caller. I wonder what it would look like if we passed in the necessary information for the APIRequest() function to figure that out for us.

Reply:
I refactored this stuff to move it to a new function `caddycmd.DetermineAdminAPIAddress`